src/hexkit/log.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Configurable logging tools with JSON output."""
+
+import json
+from collections import OrderedDict
+from datetime import datetime, timezone
+from logging import (
+    Formatter,
+    Logger,
+    LogRecord,
+    StreamHandler,
+    addLevelName,
+)
+from typing import Literal
+
+from pydantic import Field
+from pydantic_settings import BaseSettings
+
+from hexkit.correlation import (
+    correlation_id_var,
+)
+
+__all__ = [
+    "JsonFormatter",
+    "LoggingConfig",
+    "RecordCompiler",
+    "configure_logging",
+]
+
+
+DUMMY_RECORD = LogRecord("dummy", 0, "dummy", 0, None, None, None, None, None)
+RESERVED_RECORD_KEYS = DUMMY_RECORD.__dict__.keys()
+
+# Add TRACE log level
+addLevelName(5, "TRACE")
```

Comment:
Do we really need this additional level (e.g. for uvicorn) or would the default level be sufficient?

Reply:
TRACE is even more fine grained and verbose than DEBUG and normally used to get output from system calls or 3rd party libs. As far as I see it's supported by uvicorn to log on the level of ASGI events. It does not harm if we support it as well, though normally INFO or DEBUG should be sufficient to catch bugs.